src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld.java

```diff
@@ -219,16 +220,24 @@ public boolean testForCubes(CubePos start, CubePos end, Predicate<Cube> cubeAllo
     
     /**
      * @author Foghrye4
-     * @reason Original {@link World#markChunkDirty(BlockPos, TileEntity)} 
-     * called by TileEntities whenever they need to force Chunk to save 
-     * valuable info they changed.
-     * Because now we store TileEntities in Cubes instead of Chunks, it will
-     * be quite reasonable to force Cubes to save themselves.
+     * @reason Original {@link World#markChunkDirty(BlockPos, TileEntity)}
+     *         called by TileEntities whenever they need to force Chunk to save
+     *         valuable info they changed. Because now we store TileEntities in
+     *         Cubes instead of Chunks, it will be quite reasonable to force
+     *         Cubes to save themselves.
      */
     @Overwrite
     public void markChunkDirty(BlockPos pos, TileEntity unusedTileEntity) {
-        if (this.getCubeCache().getLoadedCube(CubePos.fromBlockCoords(pos)) != null) {
-            this.getCubeCache().getLoadedCube(CubePos.fromBlockCoords(pos)).markDirty();
+        if (this.isCubicWorld()) {
+            Cube cube = this.getCubeCache().getLoadedCube(CubePos.fromBlockCoords(pos));
+            if (cube != null) {
+                cube.markDirty();
+            }
+        } else {
+            Chunk chunk = this.chunkProvider.getLoadedChunk(pos.getX() >> 4, pos.getZ() >> 4);
+            if (chunk != null) {
+                chunk.setChunkModified();
+            }
```

Comment:
Instead of this, use cancellable injection. If it's cubic chunks world, cancel it and set the return value you want, otherwise don't do anything in the injection.